apps/site/assets/ts/hooks/useDepartureRow.ts

```diff
@@ -0,0 +1,99 @@
+import { useCallback, useMemo } from "react";
+import { isNil, some } from "lodash";
+import { useSearchParams } from "react-router-dom";
+import { DirectionId, Route } from "../__v3api";
+
+type DepartureRow = {
+  route: Route;
+  directionId: DirectionId;
+  headsign: string;
+};
+
+// Search params in URLs are always strings!
+type DepartureRowSearchParams = {
+  routeId: string;
+  directionId: string;
+  headsign: string;
+};
+
+const getRowParams = (
+  params: URLSearchParams
+): Record<string, string | null> => {
+  const routeId = params.get("route");
+  const directionId = params.get("direction_id");
+  const headsign = params.get("headsign");
+  return { routeId, directionId, headsign };
+};
+
+const resetRowParams = (
+  params: URLSearchParams,
+  newParams: Record<string, string | null>
+): URLSearchParams => {
+  const { routeId, directionId, headsign } = newParams;
+  Object.entries({
+    route: routeId,
+    direction_id: directionId,
+    headsign
+  }).forEach(([key, value]) => {
+    if (value) {
+      params.set(key, value);
+    } else {
+      params.delete(key);
+    }
+  });
+  return params;
+};
+
+/**
+ * Coordinates toggling the active departure row with the corresponding URL
+ * search parameters, in the format of
+ * ?route=`<routeId>`&direction_id=`<directionId>`&headsign=`<headsign>`. Needs
+ * to be called from within React Router's `<RouterProvider>` to work with URL
+ * search parameters correctly.
+ */
+const useDepartureRow = (
+  routes: Route[]
+): {
+  activeRow: DepartureRow | null;
+  resetRow: () => void;
+  setRow: (filters: DepartureRowSearchParams) => void;
+} => {
+  const [params, setSearchParams] = useSearchParams();
+
+  const activeRow = useMemo(() => {
+    const { routeId, directionId, headsign } = getRowParams(params);
+    const matchedRoute = routes.find(r => r.id === routeId);
+    if (
+      some([routeId, directionId, headsign], isNil) ||
+      !matchedRoute ||
+      !(directionId! in ["0", "1"])
```

Comment:
nit-pick(nonblocking): Instead of telling the typer that `directionId` exists we could just add a check for it `..|| !directionId ...`.  Would also be a bit clearer

Reply:
@kotva006 wouldn't that evaluate `false` when `directionId` is `0`? which is not what I'd want, since 0 is a valid value for it

Reply:
I forgot about that quirk.  Maybe [isNil](https://lodash.com/docs/4.17.15#isNil) could help here.

Reply:
yeah, I kinda went overboard and decided to do this to be strict, so if someone tries to go to an otherwise correct URL but with `direction_id=3` or something, it'll redirect to the main view instead of a broken realtime tracking view :grin: